Cannot press a key on Linux Wayland

I'm trying to simply press a key with a minimal example :

```rust
use std::thread::sleep;
use std::time::Duration;
use tfc::{traits::KeyboardContext, Context, Key};

fn main() {
    println!("Hello, world!");
    let mut ctx = Context::new().unwrap();
    loop {
        sleep(Duration::from_secs(1));
        ctx.key_click(Key::A).unwrap();
        println!("A pressed");
    }
}
```

It runs without any error but the A key isn't pressed in practice when I open a text editor for example.  I did the `chmod +0666 /dev/uinput` command but it still seems to do nothing.  Any idea of what could be wrong ?

Deleting the platform detection in `build.rs` gives a couple of compile errors:
```
error[E0308]: mismatched types
  --> The-Fat-Controller/src/command/execute.rs:62:23
   |
62 |             return Ok(true);
   |                       ^^^^ expected `()`, found `bool`

error[E0308]: mismatched types
  --> The-Fat-Controller/src/command/execute.rs:71:28
   |
71 |             _ => return Ok(false)
   |                            ^^^^^ expected `()`, found `bool`

warning: unreachable call
  --> The-Fat-Controller/src/command/execute.rs:66:9
   |
66 |            match self {
   |   _________^
   |  |_________|
   | ||
67 | ||             UnicodeCharDown(_) => panic!("UnicodeKeyboardContext is not implemented"),
68 | ||             UnicodeCharUp(_) => panic!("UnicodeKeyboardContext is not implemented"),
69 | ||             UnicodeChar(_) => panic!("UnicodeKeyboardContext is not implemented"),
70 | ||             UnicodeString(_) => panic!("UnicodeKeyboardContext is not implemented"),
71 | ||             _ => return Ok(false)
72 | ||         }?;
   | ||_________-^ unreachable call
   | |__________|
   |            any code following this `match` expression is unreachable, as all arms diverge
   |
   = note: `#[warn(unreachable_code)]` on by default

error[E0308]: mismatched types
  --> The-Fat-Controller/src/command/execute.rs:74:12
   |
74 |         Ok(true)
   |            ^^^^ expected `()`, found `bool`

error[E0308]: `?` operator has incompatible types
   --> The-Fat-Controller/src/command/execute.rs:141:12
    |
141 |         if self.execute_unicode(ctx)? {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `()`
    |
    = note: `?` operator cannot convert from `()` to `bool`

error[E0599]: no variant or associated item named `COUNT` found for enum `key::Key` in the current scope
   --> The-Fat-Controller/src/linux_wayland/mod.rs:40:33
    |
40  |         for k in 0..crate::Key::COUNT {
    |                                 ^^^^^ variant or associated item not found in `key::Key`
    |
   ::: The-Fat-Controller/src/enum.rs:110:9
    |
110 |         pub enum $name {
    |         -------------- variant or associated item `COUNT` not found here
    |
    = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
1   | use crate::r#enum::Enum;
    |

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
warning: `tfc` (lib) generated 1 warning
error: could not compile `tfc` due to 5 previous errors; 1 warning emitted
```

Not sure if the fixes I did for these were actually correct or not but the minimal example works now. I put my fixes in a fork here https://github.com/notewio/The-Fat-Controller

--- a/build.rs
+++ b/build.rs
@@ -3,7 +3,7 @@ use std::process::Command;
 fn main() {
     // Default to /dev/uinput on linux unless X11 is found
     if linux_session_type()
-        .and_then(|x| Ok(x.contains("x11")))
+        .map(|x| x.contains("x11"))
         .unwrap_or(false)
     {
         println!("cargo:rustc-cfg=x11");
@@ -17,11 +17,10 @@ fn linux_session_type() -> Result<String, Box<dyn std::error::Error>> {
     let user = std::env::var("USER")?;
     let session = std::str::from_utf8(&output.stdout)?
         .lines()
-        .filter(|x| x.contains(&user))
-        .next()
+        .find(|x| x.contains(&user))
         .unwrap_or("")
         .trim()
-        .split(" ")
+        .split(' ')
         .next()
         .unwrap_or("");
     let output = Command::new("loginctl")

--- a/build.rs
+++ b/build.rs
@@ -1,9 +1,34 @@
+use std::process::Command;
+
 fn main() {
-    // https://stackoverflow.com/a/45537237/4093378
-    if std::env::var("DISPLAY").is_ok() {
+    // Default to /dev/uinput on linux unless X11 is found
+    if linux_session_type()
+        .and_then(|x| Ok(x.contains("x11")))
+        .unwrap_or(false)
+    {
         println!("cargo:rustc-cfg=x11");
     }
+}
 
-    // There's an alternative that's more robust but also more complicated.
+fn linux_session_type() -> Result<String, Box<dyn std::error::Error>> {
+    // Really rough implementation of this:
     // https://unix.stackexchange.com/a/325972/356153
+    let output = Command::new("loginctl").output()?;
+    let user = std::env::var("USER")?;
+    let session = std::str::from_utf8(&output.stdout)?
+        .lines()
+        .filter(|x| x.contains(&user))
+        .next()
+        .unwrap_or("")
+        .trim()
+        .split(" ")
+        .next()
+        .unwrap_or("");
+    let output = Command::new("loginctl")
+        .arg("show-session")
+        .arg(session)
+        .arg("-p")
+        .arg("Type")
+        .output()?;
+    Ok(std::str::from_utf8(&output.stdout)?.to_string())
 }

--- a/src/command/execute.rs
+++ b/src/command/execute.rs
@@ -68,8 +68,8 @@ impl Command {
             UnicodeCharUp(_) => panic!("UnicodeKeyboardContext is not implemented"),
             UnicodeChar(_) => panic!("UnicodeKeyboardContext is not implemented"),
             UnicodeString(_) => panic!("UnicodeKeyboardContext is not implemented"),
-            _ => return Ok(false)
-        };
+            _ => Ok(false)
+        }
     }
 
     /// Execute a [`Command`] by calling the corresponding method on

--- a/src/command/execute.rs
+++ b/src/command/execute.rs
@@ -55,7 +55,7 @@ impl Command {
         target_os = "linux",
         not(x11)
     ))]
-    pub fn execute_unicode<C>(&self, ctx: &mut C) -> Result<(), GenericError<C::PlatformError>>
+    pub fn execute_unicode<C>(&self, ctx: &mut C) -> Result<bool, GenericError<C::PlatformError>>
         where C: FallibleContext + KeyboardContext + MouseContext + AsciiKeyboardContext
     {
         if self.execute_core(ctx)? {
@@ -69,9 +69,7 @@ impl Command {
             UnicodeChar(_) => panic!("UnicodeKeyboardContext is not implemented"),
             UnicodeString(_) => panic!("UnicodeKeyboardContext is not implemented"),
             _ => return Ok(false)
-        }?;
-
-        Ok(true)
+        };
     }
 
     /// Execute a [`Command`] by calling the corresponding method on

--- a/src/linux_wayland/mod.rs
+++ b/src/linux_wayland/mod.rs
@@ -1,11 +1,12 @@
-mod ffi;
 mod error;
+mod ffi;
 mod keyboard;
 mod mouse;
 
 // The implementation of this module is adapted from here:
 // https://www.kernel.org/doc/html/latest/input/uinput.html
 
+use crate::r#enum::Enum;
 use error::PlatformError;
 type Error = crate::GenericError<PlatformError>;
 
@@ -25,14 +26,15 @@ pub struct Context {
 
 impl Context {
     pub fn new() -> Result<Self, Error> {
-        let file = unsafe {
-            ffi::open(b"/dev/uinput\0".as_ptr(), ffi::O_WRONLY | ffi::O_NONBLOCK)
-        };
+        let file = unsafe { ffi::open(b"/dev/uinput\0".as_ptr(), ffi::O_WRONLY | ffi::O_NONBLOCK) };
         if file == -1 {
-            return Err(Error::Platform(PlatformError::errno()))
+            return Err(Error::Platform(PlatformError::errno()));
         }
 
-        let ctx = Self { file, scroll: Default::default() };
+        let ctx = Self {
+            file,
+            scroll: Default::default(),
+        };
 
         ctx.ioctl(ffi::UI_SET_EVBIT, ffi::EV_KEY)?;
         ctx.ioctl(ffi::UI_SET_EVBIT, ffi::EV_REL)?;
@@ -96,9 +98,7 @@ impl Context {
             value,
         };
         let size = std::mem::size_of::<ffi::input_event>();
-        let written = unsafe {
-            ffi::write(self.file, std::mem::transmute(&event), size)
-        };
+        let written = unsafe { ffi::write(self.file, std::mem::transmute(&event), size) };
         if written == -1 {
             Err(Error::Platform(PlatformError::errno()))
         } else if written != size as isize {

